core/src/main/java/ai/timefold/solver/core/impl/domain/variable/provided/ChangedVariableNotifier.java

```diff
@@ -0,0 +1,39 @@
+package ai.timefold.solver.core.impl.domain.variable.provided;
+
+import java.util.function.BiConsumer;
+
+import ai.timefold.solver.core.impl.domain.variable.descriptor.VariableDescriptor;
+import ai.timefold.solver.core.impl.score.director.InnerScoreDirector;
+
+public class ChangedVariableNotifier<Solution_> {
+    private final BiConsumer<VariableDescriptor<Solution_>, Object> beforeVariableChanged;
+    private final BiConsumer<VariableDescriptor<Solution_>, Object> afterVariableChanged;
+
+    public ChangedVariableNotifier(BiConsumer<VariableDescriptor<Solution_>, Object> beforeVariableChanged,
+            BiConsumer<VariableDescriptor<Solution_>, Object> afterVariableChanged) {
+        this.beforeVariableChanged = beforeVariableChanged;
+        this.afterVariableChanged = afterVariableChanged;
+    }
+
+    public static <Solution_> ChangedVariableNotifier<Solution_> empty() {
+        return new ChangedVariableNotifier<>(
+                (vd, entity) -> {
+                },
+                (vd, entity) -> {
+                });
+    }
+
+    public static <Solution_> ChangedVariableNotifier<Solution_> of(InnerScoreDirector<Solution_, ?> scoreDirector) {
```

Comment:
And this can be stored on a score director, meaning there would be a 1:1 mapping between these instances and score directors.